### DISCUSSION TITLE
Please build caffe for m1

### DISCUSSION THREAD

**Discussion by areese** [Homebrew/discussions#1459] [Everyday usage] [answered]
It looks like caffe builds and tests fine on m1

% brew install caffe --build-from-source
==> Downloading https://ghcr.io/v2/homebrew/core/glog/manifests/0.4.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/glog/blobs/sha256:8f1417f8d9c708afd66bd4b9ef1b1b1308dcd85d02950b6284ff0f712f92c5e2
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:8f1417f8d9c708afd66bd4b9ef1b1b1308dcd85d02950b6284ff0f712f92c5e2?se=2021-05-13T18%3A50%3A00Z&sig=loZ2Pp5jHSQHwEGP8q8LenW8k9ImO5DIvm9OgmNedGw%3D&s
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/lmdb/manifests/0.9.29
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/lmdb/blobs/sha256:9f0fb7ee716181ea0197b4cd2731e1a7b68077168065cf40753fdeed7ad25b70
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:9f0fb7ee716181ea0197b4cd2731e1a7b68077168065cf40753fdeed7ad25b70?se=2021-05-13T18%3A50%3A00Z&sig=HPy%2BRVgFGEsTGNKopBrurQNo%2FPA4%2BOd38Zcl762TUR
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/metis/manifests/5.1.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/metis/blobs/sha256:ea93856908a2c1c60023dd2f849339d479b20ab4ae6d51623f9496f64993ca20
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:ea93856908a2c1c60023dd2f849339d479b20ab4ae6d51623f9496f64993ca20?se=2021-05-13T18%3A50%3A00Z&sig=vIR7KDwfCsGUL7%2FzH4EUWCayDEdvPQ266AZgWfxiZyo%3D
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/tbb/manifests/2020_U3_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/tbb/blobs/sha256:33b261262efc28bb46c9aa4ff6ae74bc51033cc9f5877c3312a52af07130d534
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:33b261262efc28bb46c9aa4ff6ae74bc51033cc9f5877c3312a52af07130d534?se=2021-05-13T18%3A50%3A00Z&sig=GXzv4MUK64LNHUyRirPkVHUHPTkImx1xsDunOab2bRk%3D&s
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/suite-sparse/manifests/5.9.0-1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/suite-sparse/blobs/sha256:b45ef039ffd66c0166b487840a19c66b7ad260e32566685b61de30b3a7532719
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:b45ef039ffd66c0166b487840a19c66b7ad260e32566685b61de30b3a7532719?se=2021-05-13T18%3A50%3A00Z&sig=NYvm%2BQoeQfAbUko2ZVrc08Us8W3m4Xv506heafQadfE%3D
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/ceres-solver/manifests/2.0.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/ceres-solver/blobs/sha256:6a4b388e2ea9cf7b90477cc7ffa4b19214a93e015b10b75a395d961004dd7f67
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:6a4b388e2ea9cf7b90477cc7ffa4b19214a93e015b10b75a395d961004dd7f67?se=2021-05-13T18%3A50%3A00Z&sig=%2BI3XM7uFh0KnJICpKKkC8cvulcoElM3bJXIx7OkjYZ4%3D
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/protobuf/manifests/3.15.8
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/protobuf/blobs/sha256:a1615a95bf6f0bd3d9111fd0afa9260373295eadf147c35aef03e31abdbef6bc
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:a1615a95bf6f0bd3d9111fd0afa9260373295eadf147c35aef03e31abdbef6bc?se=2021-05-13T18%3A50%3A00Z&sig=z5dxcSizkmWOWBuNSu%2F9ZMp9q1ZQCVhVGY2UedC6uag%3D
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/double-conversion/manifests/3.1.5_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/double-conversion/blobs/sha256:928fbd4a31967ec090b4b292b1a212fec7eb75f27443493d0c175ca8bb56a9dc
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:928fbd4a31967ec090b4b292b1a212fec7eb75f27443493d0c175ca8bb56a9dc?se=2021-05-13T18%3A50%3A00Z&sig=t%2FjYOTY%2BkfFjpisXX4DqcMZjEO2IP6pIqzKnI7i0WTQ%
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/gl2ps/manifests/1.4.2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/gl2ps/blobs/sha256:02cad33d0c39773c7a0c0983f125fc04fe86d265b31cac034be45379265e65be
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:02cad33d0c39773c7a0c0983f125fc04fe86d265b31cac034be45379265e65be?se=2021-05-13T18%3A50%3A00Z&sig=lszXwHujwojNxd92SaG2pcXukRCoPay7QKuUjMt9%2F0w%3D
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/glew/manifests/2.2.0_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/glew/blobs/sha256:4ec7d501b56e5e5682f752975340c57a9aca68431d0d2cc9f849e428860f09de
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:4ec7d501b56e5e5682f752975340c57a9aca68431d0d2cc9f849e428860f09de?se=2021-05-13T18%3A50%3A00Z&sig=BqCe6Q1vAiOlQah9f3o9JnszjRto5XCqRhVpceqYY3A%3D&s
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/jsoncpp/manifests/1.9.4_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/jsoncpp/blobs/sha256:e8698be8b88f02ce485d1b787f0dfd478360861998cb14a024b13f28d89ae24d
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:e8698be8b88f02ce485d1b787f0dfd478360861998cb14a024b13f28d89ae24d?se=2021-05-13T18%3A50%3A00Z&sig=v2mAgIIU8fD05%2BkWGKAKVIsgFK2Zh%2BGbTt%2F2r9jtFk
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pugixml/manifests/1.11.4_2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pugixml/blobs/sha256:72487f35e935099779ea88bbbf6dd205406710df262674e6295972edff2a9ac1
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:72487f35e935099779ea88bbbf6dd205406710df262674e6295972edff2a9ac1?se=2021-05-13T18%3A50%3A00Z&sig=zK2NNqLvtPKnh6Bokc9pbAmsduuJgdP0VeJnjXBzDgE%3D&s
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pyqt/5/manifests/5.15.4_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pyqt/5/blobs/sha256:def5e1c4db90419439be7ac44ff3d11e6042590e3ef2e2f0586db5870b1b68fb
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:def5e1c4db90419439be7ac44ff3d11e6042590e3ef2e2f0586db5870b1b68fb?se=2021-05-13T18%3A50%3A00Z&sig=PmKTGPpkDa8UxSI9NCppKSBsgUCt6uYnGi1k%2Bek2jJU%3D
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/utf8cpp/manifests/3.2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/utf8cpp/blobs/sha256:62e5e630befe119ad905a8fa3ce2025b553309198e1fbb3b603e987d521c0eff
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:62e5e630befe119ad905a8fa3ce2025b553309198e1fbb3b603e987d521c0eff?se=2021-05-13T18%3A50%3A00Z&sig=MxnmPcYxCOW0DOAZ7SCeLZp3b5V6vPGkGFdtrO%2FtQpY%3D
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/vtk/manifests/9.0.1_9
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/vtk/blobs/sha256:825c6b028057f775e3d8d413beadf0602ca165c67571b4d94fef4a653cb648a3
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:825c6b028057f775e3d8d413beadf0602ca165c67571b4d94fef4a653cb648a3?se=2021-05-13T18%3A50%3A00Z&sig=jZuTsP6x6Ut4N7NshUO%2B%2Bt8QQZJMOHKQtqXUAywcUBE%
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/opencv/manifests/4.5.2_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/opencv/blobs/sha256:5bb57ebee8b138158d4dadef06869ccc52bdc21b5c393193db53301a90e5d9d6
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:5bb57ebee8b138158d4dadef06869ccc52bdc21b5c393193db53301a90e5d9d6?se=2021-05-13T18%3A50%3A00Z&sig=O4jpj2dFq9kIeT620TP6BTd059Vbab7qsh0pZpFHPnk%3D&s
######################################################################## 100.0%
==> Downloading https://github.com/BVLC/caffe/commit/0a04cc2ccd37ba36843c18fea2d5cbae6e7dd2b5.patch?full_index=1
######################################################################## 100.0%
==> Downloading http://dl.caffe.berkeleyvision.org/bvlc_reference_caffenet.caffemodel
######################################################################## 100.0%
==> Downloading https://github.com/BVLC/caffe/archive/1.0.tar.gz
==> Downloading from https://codeload.github.com/BVLC/caffe/tar.gz/1.0
               -=O=- #    #     #     #                                        
==> Installing dependencies for caffe: glog, lmdb, metis, tbb, suite-sparse, ceres-solver, protobuf, double-conversion, gl2ps, glew, jsoncpp, pugixml, pyqt@5, utf8cpp, vtk and opencv
==> Installing caffe dependency: glog
==> Pouring glog--0.4.0.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/glog/0.4.0: 19 files, 283.1KB
==> Installing caffe dependency: lmdb
==> Pouring lmdb--0.9.29.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/lmdb/0.9.29: 13 files, 707.7KB
==> Installing caffe dependency: metis
==> Pouring metis--5.1.0.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/metis/5.1.0: 18 files, 12.1MB
==> Installing caffe dependency: tbb
==> Pouring tbb--2020_U3_1.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/tbb/2020_U3_1: 144 files, 2.9MB
==> Installing caffe dependency: suite-sparse
==> Pouring suite-sparse--5.9.0.arm64_big_sur.bottle.1.tar.gz
🍺  /opt/homebrew/Cellar/suite-sparse/5.9.0: 163 files, 67.9MB
==> Installing caffe dependency: ceres-solver
==> Pouring ceres-solver--2.0.0.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/ceres-solver/2.0.0: 263 files, 20.4MB
==> Installing caffe dependency: protobuf
==> Pouring protobuf--3.15.8.arm64_big_sur.bottle.tar.gz
==> Caveats
Emacs Lisp files have been installed to:
  /opt/homebrew/share/emacs/site-lisp/protobuf
==> Summary
🍺  /opt/homebrew/Cellar/protobuf/3.15.8: 211 files, 17.8MB
==> Installing caffe dependency: double-conversion
==> Pouring double-conversion--3.1.5_1.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/double-conversion/3.1.5_1: 24 files, 258.3KB
==> Installing caffe dependency: gl2ps
==> Pouring gl2ps--1.4.2.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/gl2ps/1.4.2: 15 files, 331.6KB
==> Installing caffe dependency: glew
==> Pouring glew--2.2.0_1.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/glew/2.2.0_1: 38 files, 3.7MB
==> Installing caffe dependency: jsoncpp
==> Pouring jsoncpp--1.9.4_1.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/jsoncpp/1.9.4_1: 18 files, 364KB
==> Installing caffe dependency: pugixml
==> Pouring pugixml--1.11.4_2.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/pugixml/1.11.4_2: 15 files, 474.7KB
==> Installing caffe dependency: pyqt@5
==> Pouring pyqt@5--5.15.4_1.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/pyqt@5/5.15.4_1: 912 files, 27.1MB
==> Installing caffe dependency: utf8cpp
==> Pouring utf8cpp--3.2.all.bottle.tar.gz
🍺  /opt/homebrew/Cellar/utf8cpp/3.2: 11 files, 103.6KB
==> Installing caffe dependency: vtk
==> Pouring vtk--9.0.1_9.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/vtk/9.0.1_9: 3,579 files, 131.4MB
==> Installing caffe dependency: opencv
==> Pouring opencv--4.5.2_1.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/opencv/4.5.2_1: 835 files, 120.6MB
==> Installing caffe
==> Patching
==> Applying 0a04cc2ccd37ba36843c18fea2d5cbae6e7dd2b5.patch
patching file include/caffe/common.hpp
==> cmake . -DALLOW_LMDB_NOLOCK=OFF -DBUILD_SHARED_LIBS=ON -DBUILD_docs=OFF -DBUILD_matlab=OFF -DBUILD_python=OFF -DBUILD_python_layer=OFF -DCPU_ONLY=ON -DUSE_LEVELDB=ON -DUSE_LMDB=ON -DUSE_NCCL=OFF -DUSE_OPENCV=ON -DUSE_OPENMP=OF
==> make install
🍺  /opt/homebrew/Cellar/caffe/1.0_32: 156 files, 7.9MB, built in 42 seconds
==> Caveats
==> protobuf
Emacs Lisp files have been installed to:
  /opt/homebrew/share/emacs/site-lisp/protobuf

% brew test caffe
==> Testing caffe
==> /opt/homebrew/Cellar/caffe/1.0_32/bin/caffe test -model /opt/homebrew/Cellar/caffe/1.0_32/share/caffe/models/bvlc_reference_caffenet/deploy.prototxt -solver /opt/homebrew/Cellar/caffe/1.0_32/share/caffe/models/bvlc_reference_c

% brew audit --strict caffe
%

**Comment by carlocab** [comment, marked as answer]
This needs `leveldb` to be bottled first:
```
❯ brew unbottled --tag=arm64_big_sur caffe
==> Populating dependency tree...
==> :arm64_big_sur bottle status
caffe: unbottled deps: leveldb
```
Bottle job for `leveldb`: https://github.com/Homebrew/homebrew-core/actions/runs/839903254

**Reply by carlocab** [reply]
Hmm, `leveldb` failed. Probably why `caffe` hasn't been bottled yet. Probably not a difficult fix.

**Reply by carlocab** [reply]
https://github.com/Homebrew/homebrew-core/pull/77230

**Reply by areese** [reply]
Hmmm. I wonder if I had been monkeying with leveled and that's why it worked for me?

**Reply by carlocab** [reply]
You probably built it from source at some point. It builds fine, it just has the wrong `LC_RPATH` entry:
```
❯ otool -l ./leveldbutil | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 72
         path /tmp/leveldb-20210224-36583-sozoy8/leveldb-1.23/build (offset 12)
```
The PR I link to above should fix that.

**Comment by carlocab** [comment]
Bottling `leveldb` in https://github.com/Homebrew/homebrew-core/actions/runs/840239124, then will bottle `caffe`.

----

`leveldb` bottled in https://github.com/Homebrew/homebrew-core/commit/c32c610255f122e256a605cb76a5f8794679008a

`caffe` bottle job: https://github.com/Homebrew/homebrew-core/actions/runs/840250521

**Reply by carlocab** [reply]
`caffe` bottled in https://github.com/Homebrew/homebrew-core/commit/c8a5ba63606e2e52668d1a93d2285e3d03c044fb

**Reply by areese** [reply]
Thank you

I’m going through what I’ve done and looking for that rpath issue

I’ll open a discussion for the ones I find as it takes me a little longer to get the approval to send the fixes

**Comment by areese** [comment]
For the next person, this is bottled and I missed marking it answered.  :(
